File: dsa/li04_twopointers/L11_ContainerWithMostWater.java
package li04_twopointers;

public class L11_ContainerWithMostWater {

	public static void main(String[] args) {
		L11_ContainerWithMostWater obj = new L11_ContainerWithMostWater();
		System.out.println(obj.maxArea(new int[] {1,8,6,2,5,4,8,3,7}));
	}

	// Problem: https://leetcode.com/problems/container-with-most-water/
	// Idea: Move ptrs from both side. Find the max height bar from both sides and its area. Skip smaller bar. Track Max area.
	public int maxArea(int[] height) {
		int maxArea = 0;
		int l = 0, r = height.length-1;
		while(l < r) {
			int min = Math.min(height[l], height[r]);
			int area = (r-l)*min;
			maxArea =  Math.max(maxArea, area);
			if(height[l] < height[r]){
                l++;
            }else{
                r--;
            }
		}

        return maxArea;
	}

}

File: dsa/concepts/04_twopointers/python/l11_container_with_most_water.py
"""
L11 - Container With Most Water
Problem: https://leetcode.com/problems/container-with-most-water/
Idea: Move ptrs from both side. Find the max height bar from both sides and its area. Skip smaller bar. Track Max area.
Time: O(n)
Space: O(1)
"""

def max_area(height):
    """
    Find the maximum area of water that can be contained.
    
    Args:
        height: List[int] - Array representing heights of vertical lines
    
    Returns:
        int - Maximum area of water that can be contained
    """
    max_area_val = 0
    l, r = 0, len(height) - 1
    
    while l < r:
        min_height = min(height[l], height[r])
        area = (r - l) * min_height
        max_area_val = max(max_area_val, area)
        
        if height[l] < height[r]:
            l += 1
        else:
            r -= 1
    
    return max_area_val


if __name__ == "__main__":
    height = [1, 8, 6, 2, 5, 4, 8, 3, 7]
    result = max_area(height)
    print(result)  # Expected: 49
